include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+    }
+
+    bool regenerateEntryAt(std::size_t i) {
+      using namespace ::rc::detail;
+      try {
+        auto &entry = entries[i];
+        entry.shrinkable = genFunc()(entry.random, size);
+        const auto cmd = entry.shrinkable.value();
+        return true;
+      } catch (const CaseResult &result) {
+        if (result.type != CaseResult::Type::Discard) {
+          throw;
+        }
+      } catch (const GenerationFailure &failure) {
+        // Just return false below
+      }
+
+      return false;
+    }
+  };
+
+  struct ParallelCommandSequence {
+    ParallelCommandSequence(CommandSequence prefix,
+                            CommandSequence left,
+                            CommandSequence right)
+        : prefix(prefix)
+        , left(left)
+        , right(right) {}
+
+    CommandSequence prefix;
+    CommandSequence left;
+    CommandSequence right;
+  };
+
+  Shrinkable<ParallelCommands<Cmd>> generateCommands(const Random &random,
+                                                     int size) const {
+    return shrinkable::map(generateParallelSequence(random, size),
+                           [](const ParallelCommandSequence &sequence) {
+                             ParallelCommands<Cmd> cmds;
+                             copyCommandEntries(sequence.prefix, cmds.prefix);
+                             copyCommandEntries(sequence.left, cmds.left);
+                             copyCommandEntries(sequence.right, cmds.right);
+                             return cmds;
+                           });
+  }
+
+  template <typename T>
+  static void copyCommandEntries(const CommandSequence &source,
+                                 std::vector<T> &dest) {
+    const auto &entries = source.entries;
+    dest.reserve(entries.size());
+    std::transform(
+        begin(entries),
+        end(entries),
+        std::back_inserter(dest),
+        [](const CommandEntry &entry) { return entry.shrinkable.value(); });
+  }
+
+  Shrinkable<ParallelCommandSequence>
+  generateParallelSequence(const Random &random, int size) const {
+    Random r(random);
+    std::size_t count = (r.split().next() % (size + 1)) + 1;
+    return shrinkable::shrinkRecur(generateInitialParallel(random, size),
+                                   &shrinkSequence);
+  }
+
+  ParallelCommandSequence generateInitialParallel(const Random &random,
+                                                  int size) const {
+    auto r = random;
+    int prefixSz, leftSz, rightSz;
+    std::tie(prefixSz, leftSz, rightSz) = parallelCommandDistribution(size);
+
+    std::size_t prefixCount = (r.split().next() % (prefixSz + 1));
+    std::size_t leftCount = (r.split().next() % (leftSz + 1));
+    std::size_t rightCount = (r.split().next() % (rightSz + 1));
+
+    return ParallelCommandSequence(
+        generateInitial(random, prefixSz, prefixCount),
+        generateInitial(random, leftSz, leftCount),
+        generateInitial(random, rightSz, rightCount));
```

Comment:
This is super dangerous, however. These need to be split, otherwise the three sequences will get the same initial random state which means they might be identical!